packages/pyroscope-datasource-plugin/src/datasource.ts

```diff
@@ -47,32 +48,85 @@ export class DataSource extends DataSourceApi<
     return result;
   }
 
-  async metricFindQuery() {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async metricFindQuery(query: string, options?: any) {
+    const expandedQuery = getTemplateSrv().replace(query, options.scopedVars);
+    const appNamesRegex = /^\s*(apps|applications)(\(\s*\))?\s*$/;
+    const labelNamesRegex = /^\s*label_names\((\s*[\w_.-]+)\)\s*$/;
+    const labelValuesRegex =
+      /^\s*label_values\(\s*([\w_.-]+)\s*,\s*([\w_]+)\s*\)\s*$/;
+
+    const appsQuery = query.match(appNamesRegex);
+    if (appsQuery) {
+      return this.queryAppNames();
+    }
+
+    const labelNamesQuery = expandedQuery.match(labelNamesRegex);
+    if (labelNamesQuery) {
+      return this.queryLabelNames(labelNamesQuery[1]);
+    }
+
+    const labelValuesQuery = expandedQuery.match(labelValuesRegex);
+    if (labelValuesQuery) {
+      return this.queryLabelValues(labelValuesQuery[1], labelValuesQuery[2]);
+    }
+
+    return [{ text: '' }];
```

Comment:
We gotta return empty arrays, otherwise we get a single empty item:
<img width="1477" alt="Screen Shot 2023-02-24 at 20 01 39" src="https://user-images.githubusercontent.com/6951209/221281688-96eb780e-d14c-443b-9cd5-2ef1cad0df24.png">